### PR TITLE
Fix to nonlinear displaced operators

### PR DESCRIPTION
**Pull request by nantonel**
There was a bug involving displacements in expressions. 

For example using the current master and AbstractOperators `v0.0.5`:
```julia
julia> using StructuredOptimization

julia> x = Variable(1);

julia> b = 10*ones(1);

julia> @minimize ls(sigmoid(x-b))+1e-5*norm(x,1);
    it |      gamma |        fpr |        tau |        FBE |
 ------|------------|------------|------------|------------|
     1 | 7.6000e+00 | 5.6084e-02 | 1.0000e+00 | 2.6957e-02 | 
    10 | 7.6000e+00 | 9.2645e-05 | 1.0000e+00 | 9.7839e-05 | 

julia> ~x
1-element Array{Float64,1}:
 -4.57973

```

This is wrong since `b` should set the first sigmoid to be completely 0 and hence `x`, being zero, is already a local minimum.

The cause of this bug is due to the internal displacement `b` not being detected in the nonlinear operator
(when using liner operators things worked fine).
 
In fact, the displacement is merged in the function `ls` by creating a `ProximableFunction` (using `Precompose(Diagonal)`), but apparently this was done only if the displacement was present at the end of the expression e.g. the cost function:
```julia
ls( sigmoid(x) + b )
```
wouldn't be affected by this bug.  

In `v0.0.6` of `AbstractOperators` some new features were introduced:
* `AffineAdd` which is a new calculus rule that creates an affine addition to the operator (it is not named `Affine` since it should not to be confused with _affine space_ which involves linear mappings only)
* `displacement` is the function that returns the displacement of the `AffineAdd` regardless of the position where it appears in a composition of `AbstractOperators`. This should be applied only to _linear_ operators, as the displacement of the combination of nonlinear operators, contrary to linear operators, is not constant when variable changes.

I'm not going to describe the details here, but in short this is the solution that I've adopted: 
* displacements are removed from expressions and handled by AbstractOperators instead 
* when the operator is linear the displacement is still merged with the 'ProximableFunction' 
* when the operator is nonlinear the the displacement is _kept_ inside the operator 
see the [build function](https://github.com/nantonel/StructuredOptimization.jl/blob/master/src/solvers/build_solve.jl#L25). 
 
New features are included in this PR, with some new trigonometry functions and possibility of doing Hadamard products.

Finally, the problem is fixed with this PR and AbstractOperators `v0.0.6`:
```julia
julia> using StructuredOptimization

julia> x = Variable(1);

julia> b = 10*ones(1);

julia> @minimize ls(sigmoid(x-b))+1e-5*norm(x,1);
    it |      gamma |        fpr |        tau |        FBE |
 ------|------------|------------|------------|------------|
     1 | 2.3050e+08 | 0.0000e+00 | 1.0000e+00 | 1.0305e-09 | 
     1 | 2.3050e+08 | 0.0000e+00 | 1.0000e+00 | 1.0305e-09 | 

julia> ~x
1-element Array{Float64,1}:
 0.0

```

### CODE REVIEW

**Comment by lostella** [comment]
Some general comments/questions (didn't look at the code yet, correct me if any of this is wrong): should StructuredOptimization manipulate terms at all? What I mean is: right now in this PR translations (displacements, sums with constant vectors, call them as you like) are handled differently in the lienear and nonlinear case. In the linear case, an expression like
```
f(A(x + b))
```
effectively becomes 
```
f(Ax + Ab)
```
(again, correct me if I misunderstood). This is of course not doable in the nonlinear case. My reasoning here is that this should not be done in *any* case, and it should be up to the user to provide the right ("best") expression of the cost function, for one simple reason:
* If `A` is (very) tall, `A(x + b)` is much cheaper than `Ax + Ab`.

Avoiding these "manipulations" potentially has several advantages:
1. we don't need to implement any (possibly buggy) logic to attempt doing the right thing for the user => simpler code;
2. linear and nonlinear case are translated the same way => again, simpler code;
3. the user has complete control on what gets computed (and more importantly how).

It seems to me that "simply" translating an expression verbatim (as it is, without manipulations) into the functions & operators that compute it, could considerably simplify our job. But maybe there are some details that I'm missing here?

**Comment by nantonel** [comment]
Yes, initially I wanted to unify the linear and non linear operators parsing. 

I actually did that but I've noticed than that this resulted in some errors in PANOC and ZeroFPR.

I didn't investigated much and preferred into leaving the linear case untouched.

**Comment by nantonel** [comment]
Maybe we could live a further simplification on a next PR, since it would probably involve some changes in ProximalAlgorithms, although I'm not entirely sure!

File: src/syntax/terms/proximalOperators_bind.jl [hunk 1]
```diff
@@ -345,16 +345,23 @@ Equalities constraints
...
-function (==)(ex::AbstractExpression, b::Union{Real,AbstractArray}) 
-    op = operator(ex)
-    if typeof(op) <: MatrixOp
-        Term(IndAffine(op.A, b-displacement(ex), iterative = false ), variables(ex)[1])
-    else
-        # TODO change this
-        error("Currently affine equality supported only with `MatrixOp`")
-    end
-end
+# weird error!!?
```
Comment (lostella) [inline src/syntax/terms/proximalOperators_bind.jl]: What’s the problem here?

File: test/test_terms.jl [hunk 2]
```diff
@@ -162,14 +162,15 @@ cf = x == lu
...
 #IndAffine
-cf = A*x-b == 0
-@test cf.lambda == 1
-@test cf.f(~x) == (IndAffine(A,b))(~x)
-
-cf = A*x == b
-@test cf.lambda == 1
-@test cf.f(~x) == (IndAffine(A,-b))(~x)
-
+##something very weird happens here!!!
```
Comment (lostella) [inline test/test_terms.jl]: I guess commenting this is due to the other problem with the == definition that you mentioned

File: test/test_solvers.jl [hunk 3]
```diff
@@ -1,188 +0,0 @@
-@printf("\nTesting solvers\n")
```
Comment (lostella) [inline test/test_solvers.jl]: Could you just (quickly) make sure these tests are somehow in ProximalAlgorithms? If not, we should have them there at some point

Comment (nantonel) [reply test/test_solvers.jl]: yes I think we have very similar stuff in ProximalAlgorithms: 
* [normal lasso](https://github.com/kul-forbes/ProximalAlgorithms.jl/blob/master/test/test_lasso_small.jl)
*  [lasso with multiple variables](https://github.com/kul-forbes/ProximalAlgorithms.jl/blob/master/test/test_lasso_small_split_x.jl) 
* [multiple terms in the smooth functions](https://github.com/kul-forbes/ProximalAlgorithms.jl/blob/master/test/test_lasso_small_split_f.jl). 

maybe some tests like the ones of this file have to be added there as well, e.g. testing subgradients.

**Pull request merged**